From 1477508d55bd31fb88cf949b11f2d39eedbeda0a Mon Sep 17 00:00:00 2001
From: "Brian C. Lane" <bcl@redhat.com>
Date: Thu, 19 Feb 2015 10:07:17 -0800
Subject: [PATCH 5/5] libparted: device mapper uses 512b sectors (#1189328)

device mapper doesn't use the device's sector size when creating a
table. It always uses 512b units. This causes partitions to be created
8x smaller than expected on devices with 4906b sectors.

Resolves: rhbz#1189328
---
 libparted/arch/linux.c | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

diff --git a/libparted/arch/linux.c b/libparted/arch/linux.c
index 3858ab1..da7dd41 100644
--- a/libparted/arch/linux.c
+++ b/libparted/arch/linux.c
@@ -2461,8 +2461,10 @@ _dm_add_partition (PedDisk* disk, PedPartition* part)
         dm_task_destroy (task);
         task = NULL;
 
+        /* device-mapper uses 512b units, not the device's sector size */
         if (asprintf (&params, "%d:%d %lld", arch_specific->major,
-                      arch_specific->minor, part->geom.start) == -1)
+                      arch_specific->minor,
+                      part->geom.start * (disk->dev->sector_size / PED_SECTOR_SIZE_DEFAULT)) == -1)
                 goto err;
 
         task = dm_task_create (DM_DEVICE_CREATE);
@@ -2472,7 +2474,8 @@ _dm_add_partition (PedDisk* disk, PedPartition* part)
         dm_task_set_name (task, vol_name);
         if (vol_uuid)
                 dm_task_set_uuid (task, vol_uuid);
-        dm_task_add_target (task, 0, part->geom.length,
+        /* device-mapper uses 512b units, not the device's sector size */
+        dm_task_add_target (task, 0, part->geom.length * (disk->dev->sector_size / PED_SECTOR_SIZE_DEFAULT),
                 "linear", params);
         if (!dm_task_set_cookie(task, &cookie, 0))
             goto err;
-- 
2.1.0

